src/jrd/validation.cpp

```diff
@@ -1976,25 +1976,26 @@ Validation::RTN Validation::walk_index(jrd_rel* relation, index_root_page& root_
  **************************************/
 	Database* dbb = vdr_tdbb->getDatabase();
 
-	const ULONG page_number = root_page.irt_rpt[id].getRoot();
-	if (!page_number) {
+	if (root_page->irt_rpt[id].isEmpty())
 		return rtn_ok;
-	}
 
-	const bool unique = (root_page.irt_rpt[id].irt_flags & (irt_unique | idx_primary));
-	const bool descending = (root_page.irt_rpt[id].irt_flags & irt_descending);
-	const bool condition = (root_page.irt_rpt[id].irt_flags & irt_condition);
+	const ULONG page_number = root_page->irt_rpt[id].irt_root;
+	fb_assert(page_number);
+
+	const bool unique = (root_page->irt_rpt[id].irt_flags & (irt_unique | idx_primary));
+	const bool descending = (root_page->irt_rpt[id].irt_flags & irt_descending);
+	const bool condition = (root_page->irt_rpt[id].irt_flags & irt_condition);
 
-	temporary_key nullKey, *null_key = 0;
+	temporary_key nullKey, *null_key = nullptr;
 	if (unique)
 	{
 		index_desc idx;
 		{
 			// No need to evaluate index expression and/or condition
-			AutoSetRestoreFlag<UCHAR> flags(&root_page.irt_rpt[id].irt_flags,
+			AutoSetRestoreFlag<USHORT> flags(&root_page->irt_rpt[id].irt_flags,
```

Comment:
Is it necessary to specify template arg, could it be auto deducted from ctor arguments by compiler ?

Reply:
It's C++17 feature (https://en.cppreference.com/w/cpp/language/class_template_argument_deduction)
AFAIK, we were not using it, but we could.

Reply:
I've found at least one such usage of `AutoSetRestore` in remote.cpp - see `PortsCleanup::closePorts()`.

Reply:
Auto deduction works for `irt_expression`, but fails for `irt_expression | irt_condition`, it should be explicitly casted to `USHORT` to work and it somewhat defeats the idea.

Nevertheless, I'd suggest (outside this PR) to remove `AutoSetRestoreFlag` at all and instead add the needed lookup flags to `BTR_description`. Because the index root page may be locked as `LCK_read` but we modify data there (even if temporarily) -- this looks hackery.